lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+    /// Return the timestamp of the interval's end.
+    pub fn end_time(&self) -> f64 {
+        self.start_time + self.duration
+    }
+}
+
+impl Display for Measurement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str(&format!(
+            "[{:.2}, {:.2}] ({:.2}) {}",
+            self.start_time,
+            self.end_time(),
+            self.duration,
+            self.name
+        ))
+    }
+}
+
+/// Error that can occur when converting between a `PerformanceEntry` and a `Measurement`.
+#[derive(Debug, Clone)]
+pub enum MeasurementConversionError {
+    /// The name of the `PerformanceEntry` was not in a format that could be decoded to a
+    /// `Measurement`.
+    InvalidFormatting,
+    /// The log level encoded in the `PerformanceEntry`s name is not a valid log level.
+    InvalidLogLevel,
+}
+
+impl TryFrom<PerformanceEntry> for Measurement {
+    type Error = MeasurementConversionError;
+
+    fn try_from(measure: PerformanceEntry) -> Result<Self, Self::Error> {
+        use MeasurementConversionError::*;
+
+        let start_time = measure.start_time();
+        let duration = measure.duration();
+        let name = measure.name();
+        let metadata: Metadata = js_sys::Reflect::get(&measure, &"detail".to_string().into())
+            .expect("Could not get details field of PerformanceEntry")
+            .try_into()
+            .or(Err(InvalidFormatting))?;
+
+        let log_level_name = metadata.profiling_level;
+        let log_level_name = log_level_name.to_class_case();
+
+        let log_level: ProfilingLevel = from_str(&log_level_name).or(Err(InvalidLogLevel))?;
+        Ok(Measurement { start_time, duration, name, log_level })
+    }
+}
+
+/// Error that can occur when taking a measurement.
+#[derive(Clone, Debug)]
+pub enum MeasurementError {
+    /// Parsing the measurement information from the performance API failed, for example, due to an
+    /// invalid log level.
+    InvalidMeasurementConversion,
+    /// No measurement was created in the performance API backend.
+    NoMeasurementFound,
+    /// A function call to the Performance API failed to execute.
+    PerformanceAPIExecutionFailure {
+        /// Underlying error returned by the JS API.
+        error: JsValue,
+    },
+    /// Profiling for the given profiling level was disabled.
+    ProfilingDisabled,
+}
+
+
+
+// ==================================
+// === Internal Profiling Methods ===
+// ==================================
+
+/// Emit a warning if the given result is an error.
+fn warn_on_error(result: Result<Measurement, MeasurementError>) {
+    if let Err(e) = result {
+        WARNING!(format!("Failed to do profiling for an interval due to error: {:?}", e));
+    }
+}
+
+/// Return a string that encodes the given log level and name for a mark that indicates the start of
+/// an interval.
+/// Example output: "DoThing! (FilePath:22) [START]".
+fn start_interval_label(metadata: &Metadata) -> String {
```

Comment:
Is there a specific reason why these functions are not methods of `Metadata`? (`start_interval_label`, `end_interval_label`, `measure_interval_label`).